homeassistant/components/tesla_wall_connector/config_flow.py

```diff
@@ -0,0 +1,160 @@
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    wall_connector = WallConnector(
+        host=data[CONF_HOST], session=async_get_clientsession(hass)
+    )
+
+    version = await wall_connector.async_get_version()
+
+    return {
+        "title": WALLCONNECTOR_DEVICE_NAME,
+        WALLCONNECTOR_SERIAL_NUMBER: version.serial_number,
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Tesla Wall Connector."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize config flow."""
+        super().__init__()
+        self.ip_address = None
+        self.serial_number = None
+
+    async def async_step_dhcp(self, discovery_info) -> FlowResult:
+        """Handle dhcp discovery."""
+        self.ip_address = discovery_info[IP_ADDRESS]
+        _LOGGER.debug("Discovered Tesla Wall Connector at [%s]", self.ip_address)
+
+        self._async_abort_entries_match({CONF_HOST: self.ip_address})
+
+        try:
+            wall_connector = WallConnector(
+                host=self.ip_address, session=async_get_clientsession(self.hass)
+            )
+            version = await wall_connector.async_get_version()
+        except WallConnectorError as ex:
+            _LOGGER.debug(
+                "Could not read serial number from Tesla WallConnector at [%s]: [%s]",
+                self.ip_address,
+                ex,
+            )
+            return self.async_abort(reason="cannot_connect")
+
+        self.serial_number = version.serial_number
+
+        await self.async_set_unique_id(self.serial_number)
+        self._abort_if_unique_id_configured(updates={CONF_HOST: self.ip_address})
+
+        _LOGGER.debug(
+            "No entry found for wall connector with IP %s. Serial nr: %s",
+            self.ip_address,
+            self.serial_number,
+        )
+
+        placeholders = {
+            CONF_HOST: self.ip_address,
+            WALLCONNECTOR_SERIAL_NUMBER: self.serial_number,
+        }
+
+        self.context["title_placeholders"] = placeholders
+        return await self.async_step_user()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        data_schema = vol.Schema(
+            {vol.Required(CONF_HOST, default=self.ip_address): str}
+        )
+        if user_input is None:
+            return self.async_show_form(step_id="user", data_schema=data_schema)
+        errors = {}
+
+        try:
+            info = await validate_input(self.hass, user_input)
+        except WallConnectorError:
+            errors["base"] = "cannot_connect"
+        except Exception as ex:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception: %s", ex)
+            errors["base"] = "unknown"
+
+        if not errors:
+            existing_entry = await self.async_set_unique_id(
```

Comment:
We need to pass `raise_on_progress=False` here since we have discovery flows that can compete. We always want to prefer user flows.